### PR TITLE
User registration support (EZP-25730)

### PR DESCRIPTION
**Pull request by bdunogier**
> Implements [EZP-25730|https://jira.ez.no/browse/EZP-25730]

Adds real user registration support through the `/register` route. Add the `user/register` policy to the anonymous user to enable it (or use the admin account). After registering, the user is redirected to a page that confirms registration.

user_register_screenshot

### TODO
- [x] Add a custom `user/register` policy, and deny the `/register` route if the user does not have this policy
- [ ] Make target user group configurable (#88)
- [ ] Make the identifier of the Content Type used for user registration configurable
- [ ] Read user creation language from prioritized languages
- [ ] Fix the register button's label
- [ ] BDD coverage
- [ ] Improve the user account's form formatting ?

### CODE REVIEW

File: bundle/Controller/UserRegisterController.php [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+    {
+        if (!$this->isGranted(new Attribute('user', 'register'))) {
+            throw new \Exception('You are not allowed to register a new account');
+        }
+
+        $language = 'eng-GB';
+
+        $contentType = $this->repository->sudo(
+            function () {
+                return $this->contentTypeService->loadContentTypeByIdentifier('user');
```
Comment (lolautruche) [inline bundle/Controller/UserRegisterController.php]: The identifier should be configurable, like the user group

File: bundle/Controller/UserRegisterController.php [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+            $this->contentActionDispatcher->dispatchFormAction($form, $data, $form->getClickedButton()->getName());
+            if ($response = $this->contentActionDispatcher->getResponse()) {
+                return $response;
+            }
+        }
+
+        return $this->render('EzSystemsRepositoryFormsBundle:Content:content_edit.html.twig', [
+            'form' => $form->createView(),
+            'languageCode' => $language,
+            'pagelayout' => $this->pagelayout,
```
Comment (lolautruche) [inline bundle/Controller/UserRegisterController.php]: Can be null if not set by the `setPagelayout()` setter. Is this wanted?

File: lib/Form/Processor/UserRegisterFormProcessor.php [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+    /**
+     * Saves content draft corresponding to $data.
+     * Depending on the nature of $data (create or update data), the draft will either be created or simply updated.
+     *
+     * @param ContentStruct|\EzSystems\RepositoryForms\Data\User\UserCreateData $data
+     * @param $languageCode
+     *
+     * @return \eZ\Publish\API\Repository\Values\Content\Content
+     */
+    private function saveDraft(UserCreateData $data, $languageCode)
```
Comment (lolautruche) [inline lib/Form/Processor/UserRegisterFormProcessor.php]: Where is this method called?

File: lib/Form/Type/User/UserRegisterType.php [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+use Symfony\Component\OptionsResolver\OptionsResolver;
+
+/**
+ * Form type for content edition (create/update).
+ * Underlying data will be either \EzSystems\RepositoryForms\Data\Content\ContentCreateData or \EzSystems\RepositoryForms\Data\Content\ContentUpdateData
+ * depending on the context (create or update).
+ */
+class UserRegisterType extends AbstractType
+{
+    public function getName()
```
Comment (lolautruche) [inline lib/Form/Type/User/UserRegisterType.php]: Implementing this method is deprecated

File: lib/Form/Processor/UserRegisterFormProcessor.php (hunk 3 above)
Comment (bdunogier) [inline lib/Form/Processor/UserRegisterFormProcessor.php]: Line 64 above ?  :)

File: bundle/Controller/UserRegisterController.php (hunk 2 above)
Comment (bdunogier) [inline bundle/Controller/UserRegisterController.php]: Shouldn't be a problem, the templates checks it.

Otoh, maybe it doesn't make much sense not to set it.

File: bundle/Controller/UserRegisterController.php (hunk 1 above)
Comment (bdunogier) [inline bundle/Controller/UserRegisterController.php]: Todo item added.

File: lib/Form/Type/User/UserRegisterType.php (hunk 4 above)
Comment (bdunogier) [inline lib/Form/Type/User/UserRegisterType.php]: I can indeed see that, but the `FormTypeInterface` still contains it. I don't see how I could not implement it without getting myself a fatal error :)

**Comment by glye** [comment]
Re: your "not a huge fan", it is a bit weird to read. I was expecting `iDo(Not)HaveTheUserRegisterPolicy()` to login and then check the policy, but I guess the point here is not the api for reading roles, but the register action, so it does the job with a bit less code. Also re: earlier discussion about actions that change things vs assertions :) The private methods seem okay.

**Comment by bdunogier** [comment]
> iDo(Not)HaveTheUserRegisterPolicy() to login and then check the policy, but I guess the point here is not the api for reading roles

yeah, that was what I was imagining. But we don't have a way to check that using mink (we probably could with REST). Then after thinking, I concluded that it was _much_ easier that way.

About assertions/changes, it feels alright in that case, but it is a matter of perspective I guess. It is clear here that for the policies scenarii, we need to customize roles, we can't just verify, as we do not have a role with that policy in the default data. We could use the normal anonymous role for the negative check, but it will fail when we update data to allow registration by default.

Overall, creating a user with a particular role is something we do quite often. It is a bit odd to do it for the anonymous one, but our repository permits it :-)

**Comment by glye** [comment]
> scenarii

:tophat:

> In any case, i'd rather keep creating custom roles for this

To clear that up, I have nothing against creating the role. It's just that given the sentence/method naming, I was at first expecting the role to be checked at some point. Given that a) it *is* created and b) we don't care about role reading api here, it's ok :)

+1

**Comment by bdunogier** [comment]
Closing to reopen against master.

**Pull request closed without merging**